ClientLib/src/main/java/com/o3dr/android/client/utils/geotag/Geotag.java

```diff
@@ -0,0 +1,232 @@
+package com.o3dr.android.client.utils.geotag;
+
+import android.content.Context;
+import android.media.ExifInterface;
+import android.os.Build;
+import android.os.Environment;
+import android.os.Handler;
+import android.os.StatFs;
+
+import com.MAVLink.ardupilotmega.msg_camera_feedback;
+import com.o3dr.android.client.data.tlog.TLogParser;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Geotag images based on camera mavlink messages.
+ */
+public class Geotag {
+    private Geotag() {
+    }
+
+    private static ExecutorService getInstance() {
+        return InitializeExecutorService.executorService;
+    }
+
+    private static class InitializeExecutorService {
+        private static final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    }
+
+    /**
+     * Asynchronous method to geotag a list of images using a list of Events as coordinate data.
+     *
+     * Warning: this copies data to external storage
+     *
+     * @param context  {@link Context}
+     * @param handler  {@link Handler} Handler to specify what thread to callback on. This cannot be null.
+     * @param events   {@link List<com.o3dr.android.client.data.tlog.TLogParser.Event>} list of events to geotag photos.
+     * @param photos   {@link List<File>} list of files of photos to geotag.
+     * @param callback {@link Geotag.GeoTagCallback} callback to send results to.
+     */
+    public static void geoTagImagesAsync(final Context context, final Handler handler, final List<TLogParser.Event> events,
+                                         final ArrayList<File> photos, final GeoTagCallback callback) {
+
+        getInstance().execute(new Runnable() {
+            @Override
+            public void run() {
+                File saveDir = new File(getSaveRootDir(context).getPath(), "GeoTag");
+
+                if (!saveDir.exists() && !saveDir.mkdir()) {
+                    sendFailed(handler, callback, new IllegalStateException("Failed to create directory for images"));
+                    return;
+                }
+
+                GeoTagAlgorithm geoTagAlgorithm = new GeoTagAlgorithmImpl();
+                HashMap<TLogParser.Event, File> matchedPhotos = geoTagAlgorithm.match(events, photos);
+
+                if (!hasEnoughMemory(saveDir.getPath(), matchedPhotos.values())) {
```

Comment:
@chaviw `hasEnoughMemory` should take a `File` as parameter instead of a `String`.